src/libmeasurement_kit/dns/resolver.hpp

```diff
@@ -0,0 +1,152 @@
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <measurement_kit/common.hpp>
+#include <measurement_kit/dns.hpp>
+
+#include <thread>
+/// XXX use versions compatible with windows
+#include <netdb.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+namespace mk {
+namespace dns {
+
+class ResolverContext {
+  public:
+    QueryClass dns_class;
+    QueryType dns_type;
+    std::string name;
+    Callback<Error, Var<Message>> cb;
+    Settings settings;
+    Var<Reactor> reactor;
+    Var<Logger> logger;
+
+    struct addrinfo hints;
+    struct addrinfo *servinfo = nullptr;
+
+    Var<Message> message{new Message};
+
+    ResolverContext(QueryClass dns_c, QueryType dns_t, std::string n,
+                    Callback<Error, Var<Message>> c, Settings s, Var<Reactor> r,
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+
+    ~ResolverContext() {
+        if (servinfo != nullptr) {
+            freeaddrinfo(servinfo);
+        }
+    }
+};
+
+static inline void resolve_async(ResolverContext *ctx) {
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message_copy;
+    struct addrinfo *p;
+
+    int error = getaddrinfo(ctx->name.c_str(), nullptr, &(ctx->hints),
+                            &(ctx->servinfo));
+    if (error) {
+        // check the error variable and return the correct error
+        ctx->logger->warn(gai_strerror(error));
+        ctx->reactor->call_soon(
+            [callback]() { callback(ResolverError(), nullptr); });
+        delete ctx;
+        return;
+    }
+
+    if (ctx->servinfo == nullptr) {
```

Comment:
Two comments here:

1) it seems to me that `servinfo` is only needed in the scope of this function. If so, I'd restructure the code such that we can have this as a variable living on the stack (so to have minimal scope for variables)

2) my understanding is that `servinfo` must not be null after `getaddrinfo()`, is this correct? If so, probably here we can just `assert()`